### DISCUSSION TITLE
Chroma DB "Collections" - A Way To Categorize Your Documents For Meaningful Queries

### DISCUSSION THREAD

**Discussion by johnbrisbin** [imartinez/privateGPT#298] [Ideas]
I have a reasonable number of PDFs on the subject of AIs (~30). I also have a large number of fiction books leaning toward SciFi (1900+).
If I were to include them in the same privateGPT db, I would get a lot of false hits on AI queries from the SciFi corpus where AIs are a dime a dozen. What can I do?

It turns out that Chroma, the vector database currently used by privateGPT, has a convenient way of separating different sets of documents so that they can be queried separately and without confusion. This feature is called 'Collections' which is described here [Chroma - Using Collections](https://docs.trychroma.com/usage-guide#using-collections)

Collections are based on a name given when a Chroma client is created in the ingestion or query phase. Many collections can be created and each acts as if it were an entirely separate db, but they all reside in the same persist directory when forced to disk.

New collections can be added, existing ones listed, renamed or deleted. With a popup menu in a query you could select which one to use of a given query. Alternatively, a langchain program could chose which collection was most appropriate based on the nature of the query or the context of the session, similar to the way OpenAI selects plugins based on evaluation of the query.

There is some basic work needed to enable these. Obviously, this requires the ability to add new items after the initial ingestion before this can go forward. Then, added control of the collection name during ingestion and query would be required, at a minimum.

Collections will make privateGPT much more useful and effective for people who have music collections, video collections, fiction and non-fiction book collections, etc. that they want to track and query.

I would like to work with this, myself. What do you think?

**Comment by FixedAF** [comment]
keep me posted!

**Comment by FixedAF** [comment]
id offer some help but i only understand 50% of the details so far.. Short of having Vicuna or GPT4 write the code. im useless at this point, I can install and test tho! got that down.!

**Comment by johnbrisbin** [comment]
I would like to spit-ball the design issues for dealing with collections:
1. New databases start with a collection name of 'privategpt' if the user does not choose something else.
2. All existing Chroma databases will be converged to 'privategpt'. They have an initial collection when the db is created. Because we get Chroma from langchain, they set a default of 'langchain'. I would migrate the existing databases with a collection entry of 'langchain' to rename it to 'privategpt' when initially discovered, making 'privategpt' the default if the user does not choose otherwise.
3. The users preferred collection setting should be persistent until explicitly changed by the user. This implies persistent user prefs (which might be a bigger issue). In the absence of prefs and if one collection and only one collection exists, the choice is obvious. A command line setting should be supported in the interim.
4. When the user indicates (UI dependent), he should be offered a selection of all current collections and the opportunity to define a new one. The user should be able to change the current collection at any time. This appears to be low impact, but I don't know if it is effectively zero impact.
5. Additional facilities should allow the deletion of a collection (loss of ingest effort) or the renaming of a collection. Note that there are restrictions on the construction of valid names (from the Chroma docs):
Chroma uses collection names in the url, so there are a few restrictions on naming them:
- The length of the name must be between 3 and 63 characters.
- The name must start and end with a lowercase letter or a digit, and it can contain dots, dashes, and underscores in between.
- The name must not contain two consecutive dots.
- The name must not be a valid IP address.

Indexes should be logically separate among collections, but it is not clear how that is managed.

(enough for now)

**Comment by sime2408** [comment]
I build code to create a separate database for every topic I am interested in, for example, all medicine books go to the medicine database under the `db` folder, and all physics books are under the physics database sub-folder. And now, If I want to query from both databases I have to initialize twice `RetrievalQA` and ask questions twice to both databases and print the result. The problem is that Chroma receives only a single `persist_directory=persist_dir`, not a list of directories.
This frustrates me, **collections** would be very beneficial but I can't find a code on how to do it in langchain. 
Please someone help on this 🙏🏻

**Reply by johnbrisbin** [reply]
@sime2408,
There is an API described at [Chroma Docs](https://docs.trychroma.com/) .
Langchain hides unique features of plug-in components to create a lowest common denominator interface to the plug-ins. This makes it easy to substitute one for the other, but you can't get the best from any of them.
The methods from the API are tucked inside the 'db' object returned from the Chroma() call.
Methods on Client (in the API doc) appear under db._client so you would call them as  db._client.create_collection("testname")
Similarly, Methods on Collection API elements are accessed via db._collection, i.e. like db._collection.delete()

You can specify the name of the initial collection by adding the parameter to the Chroma( ) call like this:
Chroma(persist_directory=persist_directory, embedding_function=embeddings, client_settings=CHROMA_SETTINGS, collection_name="porcelain_frog_collection")

By the way, the default collection name is 'langchain'.

That pretty much exhausts what I know about it. I think it is enough to do useful things.

**Comment by sarang997** [comment]
thanks this is very useful! and on point
